### PR TITLE
LibWeb: Add a CSSPixelFraction class to allow comparison of fractions

### PR DESCRIPTION
**Pull request by Zaggy1024**
This class will allow us to compare the ratio of two `CSSPixels` values losslessly.

Not only that, but an operation like `a * (b / c)` should no longer be lossy, since the operation can be implicitly carried out as `(a * b) / c` instead. This part is demonstrated by adding parentheses to a calculation in `FormattingContext::solve_replaced_size_constraint()` to cause it to make use of the new `operator*(CSSPixels, CSSPixelFraction)`.

-------

This was created specifically to allow the changes that were initially implemented by @MacDue in the now-closed https://github.com/SerenityOS/serenity/pull/20759 to be done losslessly, rather than converting to and from floating point.

-------

Also included is a change to make `FormattingContext` use the new multiply-by-fraction mentioned above, meaning that if the compiler didn't already optimize them out, the unnecessary shifts done by the multiplication and division operators in that sequence of operations will no longer be done. This change has been verified to not break the page that was tested in https://github.com/SerenityOS/serenity/pull/20353, which would truncate if the division result was a single 32-bit value instead of a fraction.

cc @kalenikaliaksandr

### CODE REVIEW

File: Userland/Libraries/LibWeb/Layout/FormattingContext.cpp [hunk 1]
```diff
@@ -283,21 +283,21 @@ CSSPixelSize FormattingContext::solve_replaced_size_constraint(CSSPixels input_w
...
+        size = { max_width, max(min_height, max_width * (h / w)) };
     if ((w > max_width && h > max_height) && (max_width / w > max_height / h))
-        size = { max(min_width, max_height * w / h), max_height };
+        size = { max(min_width, max_height * (w / h)), max_height };
     if ((w < min_width && h < min_height) && (min_width / w <= min_height / h))
-        size = { min(max_width, min_height * w / h), min_height };
+        size = { min(max_width, min_height * (w / h)), min_height };
     if ((w < min_width && h < min_height) && (min_width / w > min_height / h))
-        size = { min_width, min(max_height, min_width * h / w) };
+        size = { min_width, min(max_height, min_width * (h / w)) };
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibWeb/Layout/FormattingContext.cpp:300]: Not a thing to do now, but this sure looks weirdly repetetive

**Review by MacDue — COMMENTED** [review]
Nit second commit remove:
> This requires doing the calculation as floats, since scaling factor may be too small to be represented as CSSPixels.

or replace with (something like):

> This makes use of the new CSSPixelFraction for the scaling factor, which avoids issues with representing small values as CSSPixels.

**Comment by Zaggy1024** [comment]
Ah, thanks, forgot to reword that when I re-committed it.

File: Userland/Libraries/LibWeb/PixelUnits.h [hunk 2]
```diff
@@ -278,6 +279,85 @@ constexpr CSSPixels operator*(unsigned long left, CSSPixels right) { return righ
 inline float operator*(float left, CSSPixels right) { return right.to_float() * left; }
 inline double operator*(double left, CSSPixels right) { return right.to_double() * left; }
 
+class CSSPixelFraction {
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/PixelUnits.h:282]: cool, I think it makes to represent temporary evaluation results using fractions to have maximum precision.

although the only thing that doesn't seem right for me is that `CSSPixelFraction` could be used outside of expression. could we make `CSSPixelFraction` to be `CSSPixels` implementation detail that should never leak outside of it?

Comment (Zaggy1024) [reply Userland/Libraries/LibWeb/PixelUnits.h:282]: I'm not sure how that would work, especially given the usage in the second commit in this PR, where we take the `min()` of multiple fractions. We need to be able to store a fraction as a variable for that to work with any reasonable amount of readability, I think.

Comment (Zaggy1024) [reply Userland/Libraries/LibWeb/PixelUnits.h:282]: It's also worth noting that with it as a public class, we should potentially be able to expand it to cover more use cases, like retaining more precision when multiplying or dividing fractions. I haven't seen a use for that yet, but it's certainly an option if any use case arises.

**Pull request merged**